### PR TITLE
Make stash support commits inside submodules

### PR DESCRIPTION
**Pull request by novalis**

### CODE REVIEW

File: node/lib/util/stash_util.js [hunk 1]
```diff
@@ -148,30 +151,159 @@ exports.save = co.wrap(function *(repo, status, includeUntracked, message) {
...
+                // meta index, and new commits in the submodule (which
+                // may or may not be related to those staged in the
+                // index).
+
+                // In theory, our data structures support writing this case,
+                // but since we don't yet support reading it, we probably
+                // shouldn't let the user get into a state that they can't
+                // easily get out of.
+
+                throw new UserError(`BUG: ${name} is in a state that is too \
```
Comment (shijinglu) [inline node/lib/util/stash_util.js:213]: probably should not call this a bug

File: node/lib/util/stash_util.js [hunk 2]
```diff
@@ -148,30 +151,159 @@ exports.save = co.wrap(function *(repo, status, includeUntracked, message) {
...
+                                          sig,
+                                          sig,
+                                          null,
+                                          "stash",
+                                          indexTree,
+                                          4,
+                                          [subCommit,
+                                           indexCommit,
+                                           indexCommit,
+                                           indexCommit]);
```
Comment (shijinglu) [inline node/lib/util/stash_util.js:197]: why does this commit have 4 parents and three of them are the same?

File: node/lib/util/stash_util.js [hunk 3]
```diff
@@ -310,30 +443,84 @@ exports.apply = co.wrap(function *(repo, id, reinstateIndex) {
...
+            } catch (e) {
+                // We expect these errors to be caught, but if
+                // something goes wrong, wWe are leaving the user in a
+                // pretty yucky state.  the alternative is to try to
+                // back out the whole stash apply, which seems worse.
+
+                rebaseError(`We are leaving the rebase half-finished, so \
+that you can fix the conflicts and continue by running 'git rebase \
+--continue' inside of ${name} (note: not 'git meta rebase').`);
+                console.error(`The underlying error, which is likely useless \
```
Comment (shijinglu) [inline node/lib/util/stash_util.js:517]: I think user usually is not interested in knowing about nodegit, would you make the message more user friendly, maybe something like "unkown error but unlikely harmful"?

File: node/test/util/stash_util.js [hunk 4]
```diff
@@ -338,6 +338,11 @@ x=E:Fmeta-stash=s;
                     }
                 }
 
+                for (const c of Object.keys(commitMap)) {
```
Comment (shijinglu) [inline node/test/util/stash_util.js:341]: why is this needed?

Comment (novalis) [reply node/test/util/stash_util.js:341]: In some cases, the first or second parent might be an existing commit, but the test framework only wants commitMap to contain commits it has never seen, so we have to remove the ones that already existed. I will add a comment to this effect.

File: node/lib/util/stash_util.js (hunk 2 above)
Comment (novalis) [reply node/lib/util/stash_util.js:197]: I'll edit the docs.

File: node/lib/util/stash_util.js (hunk 3 above)
Comment (novalis) [reply node/lib/util/stash_util.js]: (changed)

File: node/lib/util/stash_util.js (hunk 1 above)
Comment (novalis) [reply node/lib/util/stash_util.js]: (changed)

**Review by shijinglu — APPROVED** [review]
